tests/e2e/tikv/tikv.go

```diff
@@ -0,0 +1,545 @@
+				})
+				framework.ExpectNoError(err, "expect PVCs of scaled in Pods to have annotation tidb.pingcap.com/pvc-defer-deleting")
+				checkScaleInTime(scaleInTimeMap, tcase.scaleInGroups)
+			})
+		}
+	})
+
+	ginkgo.Describe("[Scale in simultaneously with asts]", func() {
+		var ocfg *tests.OperatorConfig
+		var oa *tests.OperatorActions
+		var genericCli ctrlCli.Client
+		var stsGetter typedappsv1.StatefulSetsGetter
+		var err error
+
+		ginkgo.BeforeEach(func() {
+			ocfg = e2econfig.NewDefaultOperatorConfig(cfg)
+			ocfg.Features = []string{
+				"StableScheduling=true",
+				"AdvancedStatefulSet=true",
+			}
+			oa = tests.NewOperatorActions(cli, c, asCli, aggrCli, apiExtCli, tests.DefaultPollInterval, ocfg, e2econfig.TestConfig, nil, fw, f)
+			ginkgo.By("Installing CRDs")
+			oa.CleanCRDOrDie()
+			oa.CreateCRDOrDie(ocfg)
+			ginkgo.By("Installing tidb-operator")
+			oa.CleanOperatorOrDie(ocfg)
+			oa.DeployOperatorOrDie(ocfg)
+			genericCli, err = ctrlCli.New(config, ctrlCli.Options{Scheme: scheme.Scheme})
+			framework.ExpectNoError(err, "failed to create clientset")
+			stsGetter = astsHelper.NewHijackClient(c, asCli).AppsV1()
+		})
+
+		ginkgo.AfterEach(func() {
+			ginkgo.By("Uninstall tidb-operator")
+			oa.CleanOperatorOrDie(ocfg)
+			ginkgo.By("Uninstalling CRDs")
+			oa.CleanCRDOrDie()
+		})
+
+		for _, tcase := range testCasesWithAsts {
+			tcase := tcase
+			ginkgo.It(tcase.description(), func() {
+				ginkgo.By("Deploy initial tc")
+				tc := fixture.GetTidbCluster(ns, fmt.Sprintf("scale-in-simultaneously-with-asts-%v", tcase.idx), utilimage.TiDBLatest)
+				tc.Spec.PD.Replicas = 1
+				tc.Spec.TiDB.Replicas = 1
+				tc.Spec.TiKV.Replicas = tcase.originTikvReplica
+				setDeleteSlots(tc, tcase.originDeleteSlots)
+				utiltc.MustCreateTCWithComponentsReady(genericCli, oa, tc, 5*time.Minute, 10*time.Second)
+
+				err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+					tc.Spec.TiKV.ScaleInParallelism = tests.Int32Ptr(tcase.scaleInParallelism)
+					tc.Spec.TiKV.Replicas = tcase.finalTikvReplica
+					setDeleteSlots(tc, tcase.finalDeleteSlots)
+					return nil
+				})
+				framework.ExpectNoError(err, "failed to scale in %s for TidbCluster /%s", ns, tc.Name)
+
+				var deleteSlotsList [][]int32
+				deleteSlotSets := sets.NewString()
+				ctx, cancel := context.WithCancel(context.Background())
+				ginkgo.By("Collect delete slots evolution asycn")
+				go func() {
+					for {
+						select {
+						case <-ctx.Done():
+							return
+						default:
+						}
+						sts, err := stsGetter.StatefulSets(tc.Namespace).Get(context.TODO(), controller.TiKVMemberName(tc.Name), metav1.GetOptions{})
+						if err == nil {
+							s := sts.Annotations["delete-slots"]
+							if s != "" && !deleteSlotSets.Has(s) {
+								deleteSlotSets.Insert(s)
+								var slots []int32
+								_ = json.Unmarshal([]byte(s), &slots)
+								deleteSlotsList = append(deleteSlotsList, slots)
+							}
+						}
+						time.Sleep(time.Second * 3)
```

Comment:
Is it stable? Can we not use asynchronous functions?

Reply:
I think here we can only observe the evolution of delete-slots by polling it, so sync or  async may not a key point.

But maybe the polling interval is the one can be discussed, should we make it 1 second ?

Reply:
In my opinion, asynchronous function is not easy to debug. With asynchronous function, I think we have to observe the process of this function carefully when I find any wrong evolution and want to reproduce it...

Reply:
When I was debugging #4497 which is the case that delete-slots evolution is not as expected, I used the log produced by `tidb-controller-manager` and I think is not related to the way how we collect delete-slots evolution which is just a simple for loop polling.

A sync way I can figure out is: check every delete-slots in expected from beginning to end one by one strictly and failed immediately when the expected one missed. If you think it's better I could modify the code. (or any other better sync way to check is also very nice). 

But IMO track all the delete-slots from commit spec update to final state would give more information for debugging.

Reply:
WDYT @DanielZhangQD  @csuzhangxc

Reply:
both sync or async OK for me 😂 BTW, `time.Sleep(time.Second * 3)` can be replaced with `time.After` in `select` (to replace `default`)

Reply:
Is `3 seconds` too long and may miss some middle status? What about a value like 100ms?
And please also check @csuzhangxc 's above comment.

Reply:
@lizhemingi I have discussed this comment with @KanShiori , we can check the deleteSlots update synchronously before we check `oa.WaitForTidbClusterReady` for each update, in this case, we can know exactly what happened when some error occurred the first time. However, the current solution is also good, it's up to you to change it or not.

Reply:
@DanielZhangQD Get. I will take a look but after other changes are done.